### PR TITLE
make sort kwarg in groupby more flexible

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/2762.

If we accept it then:
* by default we use the fastest algorithm
* if user passes `sort=false` explicitly then groups are created in order of appearance
* if user passes `sort=true` then groups are sorted

If we agree to add this extension I will add tests and NEWS.md entry and finalize this PR.

### CODE REVIEW

**Comment by bkamins** [comment]
@nalimilan - this PR should be good for review.

@pdeffebach - you might want to have a look at it in the context of DataFramesMeta.jl. Thank you!

File: NEWS.md [hunk 1]
```diff
@@ -1,3 +1,21 @@
...
+    `groupby` to pick the fastest available grouping algorithm;
+  - `true` sorts groups by key columns;
+  - `false` creates groups in the order of their appearance in the parent data
+    frame;
+  The old behavior was that the `sort` keyword argument allowed only `Bool` values
+  and if `false` was passed (which was the default) it corresponded to current
+  behavior when `nothing` is passed. Therefore only the user visible change is
+  when `sort=false` is passed explicitly as now instead of the order being
+  undefined it is guaranteed that the groups are in order of their appearance
+  in the parent data frame.
```
Comment (nalimilan) [inline NEWS.md:16]:
```suggestion
  In previous versions, the `sort` keyword argument allowed only `Bool` values
  and `false` (which was the default) corresponded to the new
  behavior when `nothing` is passed. Therefore only the user visible change
  affecting existing code is when `sort=false` is passed explicitly.
  The order of groups was undefined in that case, but in practice
  groups were already created in their order of appearance, *except*
  when grouping columns implemented the `DataAPI.refpool` API
  (notably `PooledArray` and `CategoricalArray`) or when they contained only
  integers in a small range.
```

File: test/grouping.jl [hunk 2]
```diff
@@ -3844,9 +3844,107 @@ end
...
+
+    df.id = PooledArray(df.id)
+    gd = groupby_checked(df, :id, sort=nothing)
+    @test Vector.(keys(gd)) == [["3"], ["1"], ["2"]]
+    gd = groupby_checked(df, :id, sort=true)
+    @test Vector.(keys(gd)) == [["1"], ["2"], ["3"]]
+    gd = groupby_checked(df, :id, sort=false)
+    @test Vector.(keys(gd)) == [["3"], ["1"], ["2"]]
+
+    # for PooledVector sorted=true is not the same as sorted=nothing
```
Comment (nalimilan) [inline test/grouping.jl:3885]:
```suggestion
    # for PooledVector sort=true is not the same as sort=nothing
```

File: test/grouping.jl [hunk 3]
```diff
@@ -3844,9 +3844,107 @@ end
...
+    df.id[1] = "300"
+    df.id[3] = "200"
+    gd = groupby_checked(df, :id, sort=nothing)
+    @test Vector.(keys(gd)) == [["1"], ["300"], ["200"]]
+    gd = groupby_checked(df, :id, sort=true)
+    @test Vector.(keys(gd)) == [["1"], ["200"], ["300"]]
+    gd = groupby_checked(df, :id, sort=false)
+    @test Vector.(keys(gd)) == [["300"], ["1"], ["200"]]
+
+    # for CategoricalVector sorted=true is the same as sorted=nothing
```
Comment (nalimilan) [inline test/grouping.jl:3895]:
```suggestion
    # for CategoricalVector sort=true is the same as sort=nothing
```

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**